client/web/src/components/views/updating-view.tsx

```diff
@@ -0,0 +1,177 @@
+  current: string,
+}) {
+  const initialUpdateState =
+    cv.RunningLatest ? UpdateState.UpToDate : UpdateState.Available
+
+  const [updateState, setUpdateState] = useState<UpdateState>(initialUpdateState)
+
+  const [updateLog, setUpdateLog] = useState<string>('')
+
+  const appendUpdateLog = (msg: string) => {
+    setUpdateLog(updateLog + msg + '\n')
+  }
+
+  useEffect(() => {
+    if (updateState !== UpdateState.Available) {
+      return () => {}
+    }
+    return installUpdate(current, setUpdateState, appendUpdateLog)
+  }, [updateState])
+
+  return (
+    <>
+      <div className="flex-1 flex flex-col justify-center items-center text-center h-full">
+        {
+          (updateState === UpdateState.InProgress) ? (
+            <>
+              <Spinner size="sm" className="text-gray-400" />
+              <h1 className="text-2xl m-3">Update in progress</h1>
+              <p className="text-gray-400">
+                The update shouldn't take more than a couple of minutes.
+                Once it's completed, you will be asked to log in again.
+              </p>
+            </>
+          ) : (updateState === UpdateState.Complete) ? (
+            <>
+              <CheckCircleIcon />
+              <h1 className="text-2xl m-3">Update complete!</h1>
+              <p className="text-gray-400">
+                You updated Tailscale{cv.LatestVersion ? ` to ${cv.LatestVersion}` : null}. <ChangelogText version={cv.LatestVersion} />
+              </p>
+              <Link
+                className="button button-blue text-sm m-3"
+                to="/"
+              >
+                Log in to access
+              </Link>
+            </>
+          ) : (updateState === UpdateState.UpToDate) ? (
+            <>
+              <CheckCircleIcon />
+              <h1 className="text-2xl m-3">Up to date!</h1>
+              <p className="text-gray-400">
+                You are already running Tailscale {current}, which is the newest version available.
+              </p>
+              <Link
+                className="button button-blue text-sm m-3"
+                to="/"
+              >
+                Return
+              </Link>
+            </>
+          ) : (
+            <>
+              <XCircleIcon />
+              <h1 className="text-2xl m-3">Update failed</h1>
+              <p className="text-gray-400">
+                Update{cv.LatestVersion ? ` to ${cv.LatestVersion}` : null} failed. TODO(naman): what now?
+              </p>
+              <Link
+                className="button button-blue text-sm m-3"
+                to="/"
+              >
+                Return
+              </Link>
+            </>
+          )
+        }
+        <pre className="h-64 overflow-scroll m-3"><code>
+          { updateLog }
+        </code></pre>
+      </div>
+    </>
+  )
+}
+
+function installUpdate(
+  currentVersion: string,
+  setUpdateState: (u: UpdateState) => void,
+  appendUpdateLog: (msg: string) => void,
```

Comment:
And then instead of passing back callbacks, you could actually create the state variables from within the hook and pass them back out to the caller component. Then the hook can update them as needed. Happy to pair later on this

Reply:
I did end up doing this, let me know if the revised approach is correct